File: interview/interview_python/binary_search.py

#二分查找算法
def binary_search(list, key):
    low, high = 0, len(list) - 1

    while low < high:
        mid = (low + high) // 2
        guess = list[mid]

        if guess == key:
            return True
        elif guess > key:
            high = mid - 1
        else:
            low = mid + 1
    return False

list = [1, 3, 6, 7, 9, 10, 12, 15]

re = binary_search(list, 3)
print(re)

re = binary_search(list, 16)
print(re)

File: interview/go/binary_search.go
package main

import "fmt"

func binarySearch(arr []int, key int) int {
	beg, end := 0, len(arr)

	for beg <= end {
		mid := (beg + end) / 2
		if arr[mid] == key {
			return mid
		} else if arr[mid] > key {
			end = mid - 1
		} else {
			beg = beg + 1
		}
	}
	return -1
}

func main() {
	arr := []int{1, 3, 4, 6, 8, 9, 12, 15, 16}
	re := binarySearch(arr, 16)
	if re != -1 {
		fmt.Println(arr[re])
	} else {
		fmt.Println("not ok")
	}
}
